pkg/registry/rbac/reconciliation/reconcile_clusterrolebindings.go

```diff
@@ -0,0 +1,250 @@
+	RoleBinding *rbac.ClusterRoleBinding
+	// Confirm indicates writes should be performed. When false, results are returned as a dry-run.
+	Confirm bool
+	// RemoveExtraSubjects indicates reconciliation should remove extra subjects from an existing role binding
+	RemoveExtraSubjects bool
+	// Client is used to look up existing rolebindings, and create/update the rolebinding when Confirm=true
+	Client internalversion.ClusterRoleBindingInterface
+}
+
+// ReconcileClusterRoleBindingResult holds the result of a reconciliation operation.
+type ReconcileClusterRoleBindingResult struct {
+	// RoleBinding is the reconciled rolebinding from the reconciliation operation.
+	// If the reconcile was performed as a dry-run, or the existing rolebinding was protected, the reconciled rolebinding is not persisted.
+	RoleBinding *rbac.ClusterRoleBinding
+
+	// MissingSubjects contains expected subjects that were missing from the currently persisted rolebinding
+	MissingSubjects []rbac.Subject
+	// ExtraSubjects contains extra subjects the currently persisted rolebinding had
+	ExtraSubjects []rbac.Subject
+
+	// Operation is the API operation required to reconcile.
+	// If no reconciliation was needed, it is set to ReconcileNone.
+	// If options.Confirm == false, the reconcile was in dry-run mode, so the operation was not performed.
+	// If result.Protected == true, the rolebinding opted out of reconciliation, so the operation was not performed.
+	// Otherwise, the operation was performed.
+	Operation ReconcileOperation
+	// Protected indicates an existing role prevented reconciliation
+	Protected bool
+}
+
+func (o *ReconcileClusterRoleBindingOptions) Run() (*ReconcileClusterRoleBindingResult, error) {
+	var result *ReconcileClusterRoleBindingResult
+
+	existingBinding, err := o.Client.Get(o.RoleBinding.Name, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		result = &ReconcileClusterRoleBindingResult{
+			RoleBinding:     o.RoleBinding,
+			MissingSubjects: o.RoleBinding.Subjects,
+			Operation:       ReconcileCreate,
+		}
+
+	case err != nil:
+		return nil, err
+
+	default:
+		result, err = computeReconciledRoleBinding(existingBinding, o.RoleBinding, o.RemoveExtraSubjects)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If reconcile-protected, short-circuit
+	if result.Protected {
+		return result, nil
+	}
+	// If we're in dry-run mode, short-circuit
+	if !o.Confirm {
+		return result, nil
+	}
+
+	switch result.Operation {
+	case ReconcileRecreate:
+		// Try deleting
+		err := o.Client.Delete(
+			existingBinding.Name,
+			&metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &existingBinding.UID}},
+		)
+		switch {
+		case err == nil, errors.IsNotFound(err):
+			// object no longer exists, as desired
+		case errors.IsConflict(err):
+			// delete failed because our UID precondition conflicted
+			// this could mean another object exists with a different UID, re-run
+			return o.Run()
```

Comment:
I think backoff is overkill here... these don't retry on update conflicts... those are handled in the rbac_storage code and already have a backoff.

I did add a max attempts counter just so we don't recurse off into the sunset if we get a degenerate case.